src/engine/sparqlExpressions/PrefilterExpressionIndex.cpp

```diff
@@ -134,39 +134,100 @@ static auto getSetUnion(const std::vector<BlockMetadata>& blocks1,
 // `BlockMetadata` values. And those `ValueId` values define the bounds of the
 // related `BlockMetadata` values.
 using RandomItPair = std::pair<RandomIt, RandomIt>;
 
+// Refers to a `BlockMetadata` span `::iterator` (index).
+using BlockSpanIt =
+    std::span<const prefilterExpressions::BlockMetadata>::iterator;
+
+//______________________________________________________________________________
+static std::pair<BlockSpanIt, BlockSpanIt> getInputAdjustedRangeIndices(
+    const BlockSpanIt& beginInput, const BlockSpanIt& endInput,
+    const RandomIt& valueIdsInputBegin, const RandomIt& firstIt,
+    const RandomIt& lastIt) {
+  return {
+      beginInput + std::distance(valueIdsInputBegin, firstIt) / 2,
+      // Round up, for Ids contained within the bounding Ids of firstTriple
+      // and lastTriple we have to include the respective metadata block
+      // (that block is partially relevant).
+      // Remark: Given lastIt refers to the highest possible input index
+      // (endInput), select endInput to avoid out-of-bound errors later on.
+      std::min(beginInput + (std::distance(valueIdsInputBegin, lastIt) + 1) / 2,
+               endInput)};
+}
+
+//______________________________________________________________________________
+// Similar to `getRelevantBlocksFromIdRanges`, but this function retrieves the
+// relevant (and save) complementing `BlockMetadata` values w.r.t. `input` and
+// `relevantIdRanges`.
+static std::vector<BlockMetadata> getComplementBlocksForRelevantIdRanges(
+    const std::vector<RandomItPair>& relevantIdRanges,
+    std::span<const BlockMetadata> input, const RandomIt& valueIdsInputBegin,
+    const RandomIt& valueIdsInputEnd) {
+  // If relevantIdRanges is empty, the complement are simply all `BlockMetadata`
+  // values (input).
+  if (relevantIdRanges.empty()) {
+    return std::vector<BlockMetadata>{input.begin(), input.end()};
+  }
+  // Use helper getInputAdjustedRangesIndices (w.r.t. std::span<const
+  // BlockMetadata> input) and bind relevant constant arguments.
+  const auto getInputAdjustedRange = std::bind(
+      getInputAdjustedRangeIndices, input.begin(), input.end(),
+      valueIdsInputBegin, std::placeholders::_1, std::placeholders::_2);
+  // Section retrieve complementing BlockMetadata values.
+  std::vector<BlockMetadata> complementBlocks;
+  complementBlocks.reserve(input.size());
+  // We implicitly handle the first complement range:
+  // input[input.begin(), (first) inputAdjustedBegin]
+  RandomIt currentBegin = valueIdsInputBegin;
+  for (const auto& [firstRelevantIt, lastRelevantIt] : relevantIdRanges) {
+    const auto& [inputAdjustedBegin, inputAdjustedEnd] =
+        getInputAdjustedRange(currentBegin, firstRelevantIt);
+    complementBlocks.insert(complementBlocks.end(), inputAdjustedBegin,
+                            inputAdjustedEnd);
+    currentBegin = lastRelevantIt;
+  }
+  // Handle the complementing range:
+  // input[(last) inputAdjustedBegin, input.end()]
```

Comment:
I'll have to think again (or you have to convince me), that this in fact works. I hope you have good unit tests for this:)

Reply:
I assume that you use a similar mechanism as for the positive case, but it is still confusing:)